File: frontend/coordinator-android/engine/src/main/kotlin/ug/qaat/engine/LecturerGate.kt
package ug.qaat.engine

import kotlin.math.ceil
import kotlin.math.max

/**
 * Lecturer gate — port of backend/api-gateway/internal/handlers/lecturer_gate_scan.go.
 * The lecturer scans the coordinator's gate QR and submits staff-ID + the live room code
 * (+ fingerprint/WebAuthn if enrolled). First valid scan = START; the second = END, which
 * only counts if a sufficient SHARE of enrolled students actually attended (anti-ghost-lecture).
 */
enum class GateState { NOT_STARTED, STARTED, ENDED }
enum class GateAction { START, END }
enum class GateRejection {
    NO_STAFF_ID, STAFF_ID_MISMATCH, BAD_ROOM_CODE, BIOMETRIC_REQUIRED, NO_QUORUM, ALREADY_ENDED,
}

data class GateResult(
    val ok: Boolean,
    val action: GateAction? = null,
    val rejection: GateRejection? = null,
    val requiredQuorum: Int = 0,
)

data class LecturerGateContext(
    val assignedStaffId: String?,     // null/blank ⇒ NO_STAFF_ID
    val roomCodeSecret: ByteArray,
    val gateState: GateState,
    val attended: Int,
    val enrolled: Int,
    val ratio: Double,                // lecturer_attendance_ratio (server default 0.5)
    val requireBiometric: Boolean,    // a WebAuthn passkey is enrolled for this lecturer
)

class LecturerGate(private val nowSeconds: () -> Long = { System.currentTimeMillis() / 1000 }) {

    fun evaluate(
        staffId: String,
        roomCode: String,
        biometricVerified: Boolean,
        ctx: LecturerGateContext,
    ): GateResult {
        val assigned = ctx.assignedStaffId?.trim().orEmpty()
        if (assigned.isEmpty()) return GateResult(false, rejection = GateRejection.NO_STAFF_ID)
        if (staffId.trim() != assigned) return GateResult(false, rejection = GateRejection.STAFF_ID_MISMATCH)

        if (!RoomCode.validate(ctx.roomCodeSecret, roomCode.trim(), nowSeconds()))
            return GateResult(false, rejection = GateRejection.BAD_ROOM_CODE)

        if (ctx.requireBiometric && !biometricVerified)
            return GateResult(false, rejection = GateRejection.BIOMETRIC_REQUIRED)

        return when (ctx.gateState) {
            GateState.NOT_STARTED -> GateResult(true, action = GateAction.START)
            GateState.ENDED -> GateResult(false, rejection = GateRejection.ALREADY_ENDED)
            GateState.STARTED -> {
                val required = max(1, ceil(ctx.ratio * ctx.enrolled).toInt())
                if (ctx.attended < required)
                    GateResult(false, rejection = GateRejection.NO_QUORUM, requiredQuorum = required)
                else
                    GateResult(true, action = GateAction.END, requiredQuorum = required)
            }
        }
    }
}

File: frontend/coordinator-flutter/lib/engine/lecturer_gate.dart
import 'dart:math';

import 'room_code.dart' show RoomCode;

enum GateState { notStarted, started, ended }

enum GateAction { start, end }

enum GateRejection {
  noStaffId,
  staffIdMismatch,
  badRoomCode,
  biometricRequired,
  noQuorum,
  alreadyEnded,
}

class GateResult {
  final bool ok;
  final GateAction? action;
  final GateRejection? rejection;
  final int requiredQuorum;

  const GateResult(this.ok, {this.action, this.rejection, this.requiredQuorum = 0});
}

class LecturerGateContext {
  final String? assignedStaffId;
  final List<int> roomCodeSecret;
  final GateState gateState;
  final int attended;
  final int enrolled;
  final double ratio;
  final bool requireBiometric;

  const LecturerGateContext({
    this.assignedStaffId,
    required this.roomCodeSecret,
    required this.gateState,
    required this.attended,
    required this.enrolled,
    required this.ratio,
    required this.requireBiometric,
  });
}

class LecturerGate {
  final int Function() nowSeconds;

  LecturerGate({int Function()? nowSeconds})
      : nowSeconds = nowSeconds ?? _defaultNowSeconds;

  static int _defaultNowSeconds() => DateTime.now().millisecondsSinceEpoch ~/ 1000;

  GateResult evaluate(
    String staffId,
    String roomCode,
    bool biometricVerified,
    LecturerGateContext ctx,
  ) {
    final assigned = ctx.assignedStaffId?.trim() ?? '';
    if (assigned.isEmpty) {
      return const GateResult(false, rejection: GateRejection.noStaffId);
    }
    if (staffId.trim() != assigned) {
      return const GateResult(false, rejection: GateRejection.staffIdMismatch);
    }

    if (!RoomCode.validate(ctx.roomCodeSecret, roomCode.trim(), nowSeconds())) {
      return const GateResult(false, rejection: GateRejection.badRoomCode);
    }

    if (ctx.requireBiometric && !biometricVerified) {
      return const GateResult(false, rejection: GateRejection.biometricRequired);
    }

    switch (ctx.gateState) {
      case GateState.notStarted:
        return const GateResult(true, action: GateAction.start);
      case GateState.ended:
        return const GateResult(false, rejection: GateRejection.alreadyEnded);
      case GateState.started:
        final required = max(1, (ctx.ratio * ctx.enrolled).ceil());
        if (ctx.attended < required) {
          return GateResult(false, rejection: GateRejection.noQuorum, requiredQuorum: required);
        }
        return GateResult(true, action: GateAction.end, requiredQuorum: required);
    }
  }
}